## Project Description 

### AIM

## Introduction to Data Science Job

Tasks -
1. Basic data cleaning and feature exploration
2. Exploratory data analysis
3. Pipelines
4. Model experimentation 
5. Feature Engineering

In [ ]:
# Import the relevant packages
import pandas as pd
import numpy as np
import plotly.graph_objects as go
import plotly.express as px
import scipy.stats as stats

### Basic data exploration
1. check the datasets
2. combine the datasets
3. Look at summary statistics
4. Clean the dataset
5. Check for the null values


In [8]:
# get data
job2 = pd.read_csv('AustraliaDataScienceJob2.csv')
job = pd.read_csv('AustraliaDataScienceJobs.csv')
print(f'job2 dataset shape = {job2.shape}')
print(f'job dataset shape = {job.shape}')

job2 dataset shape = (652, 53)
job dataset shape = (2088, 53)


In [9]:
# combine both the dataset
job = pd.concat([job, job2])
print(f'Shape of combined dataset = {job.shape} ')

Shape of combined dataset = (2740, 53) 


In [10]:
job.head(5)

Job Title Job Location               Company  \
0                      Analyst    Melbourne     ANZ Banking Group   
1  Clinical Research Associate     Mulgrave  Bristol Myers Squibb   
2  Clinical Research Associate     Mulgrave  Bristol Myers Squibb   
3  Clinical Research Associate     Mulgrave  Bristol Myers Squibb   
4               Data Scientist    Melbourne     ANZ Banking Group   

                                                 Url  Estimate Base Salary  \
0  https://www.glassdoor.com.au/partner/jobListin...                 95917   
1  https://www.glassdoor.com.au/partner/jobListin...                 96555   
2  https://www.glassdoor.com.au/partner/jobListin...                 96555   
3  https://www.glassdoor.com.au/partner/jobListin...                 96555   
4  https://www.glassdoor.com.au/partner/jobListin...                115631   

   Low Estimate  High Estimate      Company Size      Company Type  \
0         80000         115000  10000+ Employees  Company - Public   
1         79000         118000  10000+ Employees  Company - Public   
2         79000         118000  10000+ Employees  Company - Public   
3         79000         118000  10000+ Employees  Company - Public   
4         94000         143000  10000+ Employees  Company - Public   

                   Company Sector  ...  cassandra_yn hive_yn bigml_yn  \
0                         Finance  ...             0       0        0   
1  Pharmaceutical & Biotechnology  ...             0       0        0   
2  Pharmaceutical & Biotechnology  ...             0       0        0   
3  Pharmaceutical & Biotechnology  ...             0       0        0   
4                         Finance  ...             0       0        0   

  tableau_yn  powerbi_yn  nlp_yn  pytorch_yn  tensorflow_yn  mathematic_yn  \
0          0           0       0           0              0              1   
1          0           0       0           0              0              0   
2          0           0       0           0              0              0   
3          0           0       0           0              0              0   
4          0           0       0           0              0              0   

   statistic_yn  
0             0  
1             0  
2             0  
3             0  
4             0  

[5 rows x 53 columns]

In [11]:
job.tail(5)

Job Title       Job Location                        Company  \
647    Electronics Engineer  Western Australia  Australian Antarctic Division   
648            Construction              Perth                  Chaleen Botha   
649            Construction              Perth                  Chaleen Botha   
650  Metallurgical Engineer              Perth                            BHP   
651  Metallurgical Engineer              Perth                            BHP   

                                                   Url  Estimate Base Salary  \
647  https://www.glassdoor.com.au/partner/jobListin...                 92076   
648  https://www.glassdoor.com.au/partner/jobListin...                 55000   
649  https://www.glassdoor.com.au/partner/jobListin...                 55000   
650  https://www.glassdoor.com.au/partner/jobListin...                141443   
651  https://www.glassdoor.com.au/partner/jobListin...                141443   

     Low Estimate  High Estimate           Company Size      Company Type  \
647         92000          92000  501 to 1000 Employees        Government   
648         50000          60000                    NaN               NaN   
649         50000          60000                    NaN               NaN   
650        121000         165000       10000+ Employees  Company - Public   
651        121000         165000       10000+ Employees  Company - Public   

                         Company Sector  ...  cassandra_yn hive_yn bigml_yn  \
647  Government & Public Administration  ...             0       0        0   
648                                 NaN  ...             0       0        0   
649                                 NaN  ...             0       0        0   
650           Energy, Mining, Utilities  ...             0       0        0   
651           Energy, Mining, Utilities  ...             0       0        0   

    tableau_yn  powerbi_yn  nlp_yn  pytorch_yn  tensorflow_yn  mathematic_yn  \
647          0           0       0           0              0              0   
648          0           0       0           0              0              0   
649          0           0       0           0              0              0   
650          0           0       0           0              0              0   
651          0           0       0           0              0              0   

     statistic_yn  
647             0  
648             0  
649             0  
650             0  
651             0  

[5 rows x 53 columns]

In [12]:
job.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2740 entries, 0 to 651
Data columns (total 53 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Job Title                      2737 non-null   object 
 1   Job Location                   2740 non-null   object 
 2   Company                        2740 non-null   object 
 3   Url                            2740 non-null   object 
 4   Estimate Base Salary           2740 non-null   int64  
 5   Low Estimate                   2740 non-null   int64  
 6   High Estimate                  2740 non-null   int64  
 7   Company Size                   2469 non-null   object 
 8   Company Type                   2469 non-null   object 
 9   Company Sector                 2000 non-null   object 
 10  Company Founded                1569 non-null   float64
 11  Company Industry               2000 non-null   object 
 12  Company Revenue                2469 non-null   ob

Depict - 
1. There are 2740 entries in the dataset, which means that it is fairly small by machine learning standards.
2. Null values in 
    1. Job Title
    2. Company Size
    3. Company Type
    4. Company Secto
    5. Company Founded
    6. Company Industry
    7. Company Revenue
    8. Job Descriptions
    9. Company Rating
    10. Company Friend Reccomendation
    11. Company CEO Approval           
    12. Companny Number of Rater       
    13. Company Career Opportinities   
    14. Compensation and Benefits      
    15. Company Culture and Values     
    16. Company Senior Management      
    17. Company Work Life Balance


In [15]:
# Summary of categorical features
cat_feature = job.select_dtypes(include=[object])
cat_feature.describe().T

count unique  \
Job Title         2737    352   
Job Location      2740    117   
Company           2740    485   
Url               2740   2740   
Company Size      2469      8   
Company Type      2469     12   
Company Sector    2000     24   
Company Industry  2000     65   
Company Revenue   2469     14   
Job Descriptions  2739    807   
Country           2740      1   
State             2740      7   

                                                                top  freq  
Job Title                                            Data Scientist   449  
Job Location                                              Melbourne   770  
Company                                                    Deloitte   144  
Url               https://www.glassdoor.com.au/partner/jobListin...     1  
Company Size                                       10000+ Employees   831  
Company Type                                       Company - Public   889  
Company Sector                                              Finance   401  
Company Industry                       National Services & Agencies   223  
Company Revenue                            Unknown / Non-Applicable   968  
Job Descriptions  CSIRO’s Data61 is hiring:\nWe have a range of ...    40  
Country                                                   Australia  2740  
State                                                      Victoria  1050

#### Depict
1. The URL feature comprises entirely distinct values, which means it will not have any impact on the prediction
2. The country feature has only one unique value, therefore it will not contribute to the prediction.

In [18]:
# Summary of numerical features
num_features = job.select_dtypes(include=[np.number])
num_features.describe().T

count           mean           std      min  \
Estimate Base Salary           2740.0  103185.945255  32620.513310  40500.0   
Low Estimate                   2740.0   91380.291971  29172.947652  30000.0   
High Estimate                  2740.0  116798.905109  38793.121065  41000.0   
Company Founded                1569.0    1942.167623     63.627829   1631.0   
Company Rating                 2314.0       3.884313      0.517118      1.6   
Company Friend Reccomendation  2225.0      76.335281     14.861318     15.0   
Company CEO Approval           1749.0      77.934248     16.725428      7.0   
Companny Number of Rater       1892.0    1757.846723   4647.614482      0.0   
Company Career Opportinities   2270.0       3.690529      0.542892      1.0   
Compensation and Benefits      2270.0       3.628502      0.531140      1.0   
Company Culture and Values     2270.0       3.852335      0.564251      1.0   
Company Senior Management      2270.0       3.452819      0.575394      1.0   
Company Work Life Balance      2270.0       3.772291      0.589988      1.0   
python_yn                      2740.0       0.325912      0.468800      0.0   
r_yn                           2740.0       0.180292      0.384501      0.0   
sql_yn                         2740.0       0.327737      0.469474      0.0   
java_yn                        2740.0       0.091606      0.288521      0.0   
julia_yn                       2740.0       0.002920      0.053965      0.0   
scala_yn                       2740.0       0.041606      0.199723      0.0   
c_yn                           2740.0       0.017153      0.129866      0.0   
c++_yn                         2740.0       0.076277      0.265490      0.0   
javascript_yn                  2740.0       0.024818      0.155597      0.0   
spark_yn                       2740.0       0.054015      0.226088      0.0   
hadoop_yn                      2740.0       0.024818      0.155597      0.0   
matlab_yn                      2740.0       0.014964      0.121429      0.0   
sas_yn                         2740.0       0.024818      0.155597      0.0   
git_yn                         2740.0       0.048540      0.214944      0.0   
excel_yn                       2740.0       0.126642      0.332633      0.0   
nosql_yn                       2740.0       0.004380      0.066045      0.0   
aws_yn                         2740.0       0.096350      0.295125      0.0   
mongodb_yn                     2740.0       0.012044      0.109101      0.0   
cassandra_yn                   2740.0       0.000365      0.019104      0.0   
hive_yn                        2740.0       0.039416      0.194618      0.0   
bigml_yn                       2740.0       0.000000      0.000000      0.0   
tableau_yn                     2740.0       0.090876      0.287485      0.0   
powerbi_yn                     2740.0       0.049635      0.217229      0.0   
nlp_yn                         2740.0       0.044891      0.207101      0.0   
pytorch_yn                     2740.0       0.009124      0.095101      0.0   
tensorflow_yn                  2740.0       0.014234      0.118474      0.0   
mathematic_yn                  2740.0       0.193431      0.395060      0.0   
statistic_yn                   2740.0       0.144891      0.352055      0.0   

                                   25%       50%       75%       max  
Estimate Base Salary           80623.0   96130.0  120181.0  295000.0  
Low Estimate                   73000.0   87000.0  106000.0  241000.0  
High Estimate                  90000.0  114000.0  140000.0  349000.0  
Company Founded                 1888.0    1949.0    2001.0    2020.0  
Company Rating                     3.7       3.9       4.1       5.0  
Company Friend Reccomendation     69.0      79.0      85.0     100.0  
Company CEO Approval              72.0      83.0      88.0     100.0  
Companny Number of Rater           0.0      42.0    1014.0   30783.0  
Company Career Opportinities       3.4       3.7       4.0       5.0  
Com